### ISSUE TITLE
feature request: new rule for invalid React child in JSX

### ISSUE THREAD

**Issue by jer-sen** [typescript-eslint/typescript-eslint#3224]
- [x ] I have tried restarting my IDE and the issue persists.
- [x ] I have updated to the latest version of the packages.
- [x ] I have [read the FAQ](https://github.com/typescript-eslint/typescript-eslint/blob/master/docs/getting-started/linting/FAQ.md) and my problem is not listed.

**Repro**

```JSON
{
  "rules": {
  }
}
```

```TS
import * as React from 'react';

const dangerousJsx = <>{{ objectProp: true }}</>;

```

**Expected Result**

An eslint warning to avoid runtime error such as `Objects are not valid as a React child (found: object with keys {XXX}). If you meant to render a collection of children, use an array instead.` if `dangerousJsx` is rendered.

**Actual Result**

No way to get a warning.

**Additional Info**

This is a common issue cf these example of errors that could have been avoided:
- https://github.com/yannickcr/eslint-plugin-react/issues/1293
- https://stackoverflow.com/questions/59817384/objects-are-not-valid-as-a-react-child
- https://stackoverflow.com/questions/62287705/react-typescript-objects-are-not-valid-as-a-react-child

And people are looking for such a rule:
- https://stackoverflow.com/questions/64992163/are-there-any-eslint-rules-that-could-prevent-objects-are-not-valid-as-a-react
- https://github.com/DefinitelyTyped/DefinitelyTyped/issues/35622
- https://github.com/DefinitelyTyped/DefinitelyTyped/issues/49705
- https://github.com/DefinitelyTyped/DefinitelyTyped/issues/34530
- https://github.com/DefinitelyTyped/DefinitelyTyped/issues/29307
- https://github.com/DefinitelyTyped/DefinitelyTyped/pull/33346/files

Note that a strict type check is not possible since this is valid with JSX to use anything as a node. It's just dangerous because if at the end React get this JSX to render, a runtime error will occur. That's why the warning must come from the linter.

**Versions**

| package                            | version |
| ---------------------------------- | ------- |
| `@typescript-eslint/eslint-plugin` | `4.18.0` |
| `@typescript-eslint/parser`        | `4.18.0` |
| `TypeScript`                       | `4.2.3` |
| `ESLint`                           | `7.22.0` |
| `node`                             | `14.15.5` |

**Comment by bradzacher** [comment]
A few things I can see with a quick scan of the issues.

First of all this isn't a linting issue, it's a typing issue. The types are wrong, but fixable - someone just needs to fix them.

Second we don't want to encode react-specific semantics into any rule within this plugin.
Type aware linting isn't something restricted to this project - any plugin can implement type aware rules (there are many out there that do).

Finally there are many cases you just can't catch - even with correct types.
For example `<>{() => 'foo'}</>` is 100% valid and correct code - linting against it would be wrong.

**Comment by jer-sen** [comment]
Ok
